##  Credit Scoring System

###  Business Understanding

####  Objective
The goal of this project is to build a Credit Scoring system that classifies customers into **Good**, **Average**, or **Poor** credit categories based on their financial and behavioral data.

#### Requirements & Expectations

- Understand the key business questions and challenges related to credit risk management:
  - Who are high-risk customers likely to default on loans or payments?
  - What factors most influence the credit score?
  - How can the credit scoring model help banks reduce losses and approve credit more accurately?

- Deliverables:
  - A cleaned and preprocessed dataset ready for modeling
  - A predictive model for credit scoring with evaluated performance metrics
  - Visualizations to help interpret customer credit behaviors
  - A user-friendly interface or report for displaying credit score results

- Expected Outcome:
  - A reliable credit scoring solution that supports decision-making in financial institutions
  - Improved accuracy in identifying customers with potential payment delays or defaults

---

####  Data Overview

The dataset contains customer financial and behavioral information:

| Feature                | Description                                               |
|------------------------|-----------------------------------------------------------|
| ID                     | Unique ID of the record                                   |
| Customer_ID            | Unique ID of the customer                                 |
| Month                  | Month of the year                                        |
| Name                   | The name of the person                                   |
| Age                    | The age of the person                                    |
| SSN                    | Social Security Number                                   |
| Occupation             | The occupation of the person                             |
| Annual_Income          | The annual income                                       |
| Monthly_Inhand_Salary  | Monthly in-hand salary                                  |
| Num_Bank_Accounts      | Number of bank accounts                                  |
| Num_Credit_Card        | Number of credit cards                                  |
| Interest_Rate          | Interest rate on credit card                            |
| Num_of_Loan            | Number of loans taken                                   |
| Type_of_Loan           | Types of loans                                         |
| Delay_from_due_date    | Average days delayed from due date                      |
| Num_of_Delayed_Payment | Number of payments delayed                              |
| Changed_Credit_Card    | Percentage change in credit card limit                  |
| Num_Credit_Inquiries   | Number of credit card inquiries                         |
| Credit_Mix             | Classification of credit mix                            |
| Outstanding_Debt       | Outstanding debt                                       |
| Credit_Utilization_Ratio | Credit utilization ratio                              |
| Credit_History_Age     | Age of credit history                                 |
| Payment_of_Min_Amount  | Whether minimum amount was paid (Yes/No)                |
| Total_EMI_per_month    | Total EMI per month                                     |
| Amount_invested_monthly | Monthly amount invested                                 |
| Payment_Behaviour      | Payment behaviour                                      |
| Monthly_Balance        | Monthly balance left                                   |
| Credit_Score           | The credit score (target variable)                      |

---

### Installation
- numpy
- pandas
- matplotlib
- seaborn
- sklearn

# Thư viện

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv('/content/credit.csv')
df.head()

ID  Customer_ID  Month           Name   Age          SSN Occupation  \
0  5634         3392      1  Aaron Maashoh  23.0  821000265.0  Scientist   
1  5635         3392      2  Aaron Maashoh  23.0  821000265.0  Scientist   
2  5636         3392      3  Aaron Maashoh  23.0  821000265.0  Scientist   
3  5637         3392      4  Aaron Maashoh  23.0  821000265.0  Scientist   
4  5638         3392      5  Aaron Maashoh  23.0  821000265.0  Scientist   

   Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  ...  Credit_Mix  \
0       19114.12            1824.843333                3.0  ...        Good   
1       19114.12            1824.843333                3.0  ...        Good   
2       19114.12            1824.843333                3.0  ...        Good   
3       19114.12            1824.843333                3.0  ...        Good   
4       19114.12            1824.843333                3.0  ...        Good   

   Outstanding_Debt  Credit_Utilization_Ratio Credit_History_Age  \
0            809.98                 26.822620              265.0   
1            809.98                 31.944960              266.0   
2            809.98                 28.609352              267.0   
3            809.98                 31.377862              268.0   
4            809.98                 24.797347              269.0   

   Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                     No            49.574949                 21.46538   
1                     No            49.574949                 21.46538   
2                     No            49.574949                 21.46538   
3                     No            49.574949                 21.46538   
4                     No            49.574949                 21.46538   

                  Payment_Behaviour Monthly_Balance  Credit_Score  
0   High_spent_Small_value_payments      312.494089          Good  
1    Low_spent_Large_value_payments      284.629162          Good  
2   Low_spent_Medium_value_payments      331.209863          Good  
3    Low_spent_Small_value_payments      223.451310          Good  
4  High_spent_Medium_value_payments      341.489231          Good  

[5 rows x 28 columns]

# Hiểu dữ liệu

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  int64  
 1   Customer_ID               100000 non-null  int64  
 2   Month                     100000 non-null  int64  
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  float64
 5   SSN                       100000 non-null  float64
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  float64
 8   Monthly_Inhand_Salary     100000 non-null  float64
 9   Num_Bank_Accounts         100000 non-null  float64
 10  Num_Credit_Card           100000 non-null  float64
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  float64
 13  Type_of_Loan              100000 non-null  ob

- Bộ dữ liệu gồm có 100000 bản ghi và 28 đặc trưng, mỗi bản ghi tương ứng với 1 khách hàng

In [6]:
# thống kê:
df.describe()

ID    Customer_ID          Month            Age  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean    80631.500000   25982.666640       4.500000      33.316340   
std     43301.486619   14340.543051       2.291299      10.764812   
min      5634.000000    1006.000000       1.000000      14.000000   
25%     43132.750000   13664.500000       2.750000      24.000000   
50%     80631.500000   25777.000000       4.500000      33.000000   
75%    118130.250000   38385.000000       6.250000      42.000000   
max    155629.000000   50999.000000       8.000000      56.000000   

                SSN  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
count  1.000000e+05  100000.000000          100000.000000      100000.000000   
mean   5.004617e+08   50505.123449            4197.270835           5.368820   
std    2.908267e+08   38299.422093            3186.432497           2.593314   
min    8.134900e+04    7005.930000             303.645417           0.000000   
25%    2.451686e+08   19342.972500            1626.594167           3.000000   
50%    5.006886e+08   36999.705000            3095.905000           5.000000   
75%    7.560027e+08   71683.470000            5957.715000           7.000000   
max    9.999934e+08  179987.280000           15204.633333          11.000000   

       Num_Credit_Card  Interest_Rate  ...  Delay_from_due_date  \
count    100000.000000   100000.00000  ...         100000.00000   
mean          5.533570       14.53208  ...             21.08141   
std           2.067098        8.74133  ...             14.80456   
min           0.000000        1.00000  ...              0.00000   
25%           4.000000        7.00000  ...             10.00000   
50%           5.000000       13.00000  ...             18.00000   
75%           7.000000       20.00000  ...             28.00000   
max          11.000000       34.00000  ...             62.00000   

       Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
count           100000.000000         100000.000000         100000.000000   
mean                13.313120             10.470323              5.798250   
std                  6.237166              6.609481              3.867826   
min                  0.000000              0.500000              0.000000   
25%                  9.000000              5.380000              3.000000   
50%                 14.000000              9.400000              5.000000   
75%                 18.000000             14.850000              8.000000   
max                 25.000000             29.980000             17.000000   

       Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
count     100000.000000             100000.000000       100000.000000   
mean        1426.220376                 32.285173          221.220460   
std         1155.129026                  5.116875           99.680716   
min            0.230000                 20.000000            1.000000   
25%          566.072500                 28.052567          144.000000   
50%         1166.155000                 32.305784          219.000000   
75%         1945.962500                 36.496663          302.000000   
max         4998.070000                 50.000000          404.000000   

       Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance  
count        100000.000000            100000.000000    100000.000000  
mean            107.699208                55.101315       392.697586  
std             132.267056                39.006932       201.652719  
min               0.000000                 0.000000         0.007760  
25%              29.268886                27.959111       267.615983  
50%              66.462304                45.156550       333.865366  
75%             147.392573                71.295797       463.215683  
max            1779.103254               434.191089      1183.930696  

[8 rows x 21 columns]

- độ tuổi trung bình của khách hàng là 33 tuổi; thấp nhất là 14 tuổi; cao nhất 56 tuổi
- thu nhập trung bình hằng năm của khách hàng là 50505 đô; thấp nhất là 7005 đô; cao nhất là 179987 đô

In [7]:
# thống kê dữ liệu định tính:
df.describe(include='object')

Name Occupation Type_of_Loan Credit_Mix Payment_of_Min_Amount  \
count     100000     100000       100000     100000                100000   
unique     10128         15         6261          3                     3   
top     Jessicad     Lawyer      No Data   Standard                   Yes   
freq          48       7096        11408      45848                 52326   

                     Payment_Behaviour Credit_Score  
count                           100000       100000  
unique                               6            3  
top     Low_spent_Small_value_payments     Standard  
freq                             28616        53174

- có 10128 đối tượng khách hàng
- điểm tín dụng(Credit_Score) có 3 mức
- mức điểm tín dụng chiếm nhiều nhất là: Standard(tiêu chuẩn) với số lượng 53174 bản ghi


In [9]:
df["Credit_Mix"].unique()

array(['Good', 'Standard', 'Bad'], dtype=object)

In [8]:
# đếm số Standard tại cột Credit_Score:
df['Credit_Score'].value_counts()

Credit_Score
Standard    53174
Poor        28998
Good        17828
Name: count, dtype: int64

# TIỀN XỬ LÝ DỮ LIỆU

# kiểm tra missing data

In [5]:
df.isnull().sum()

ID                          0
Customer_ID                 0
Month                       0
Name                        0
Age                         0
SSN                         0
Occupation                  0
Annual_Income               0
Monthly_Inhand_Salary       0
Num_Bank_Accounts           0
Num_Credit_Card             0
Interest_Rate               0
Num_of_Loan                 0
Type_of_Loan                0
Delay_from_due_date         0
Num_of_Delayed_Payment      0
Changed_Credit_Limit        0
Num_Credit_Inquiries        0
Credit_Mix                  0
Outstanding_Debt            0
Credit_Utilization_Ratio    0
Credit_History_Age          0
Payment_of_Min_Amount       0
Total_EMI_per_month         0
Amount_invested_monthly     0
Payment_Behaviour           0
Monthly_Balance             0
Credit_Score                0
dtype: int64

- bộ dữ liệu không có giá trị thiếu

# giải quyết dữ liệu không nhất quán